### PR TITLE
[DCA] Query CLC Runners stats

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

- adds a http client to be used by the DCA to query CLC Runners stats
- updates DCA cache with CLC Runners stats

### Motivation

Rebalance and optimize the cluster level checks dispatching logic

### Additional Notes

https://github.com/DataDog/datadog-agent/pull/3931

### CODE REVIEW

File: cmd/cluster-agent/api/v1/clusterchecks.go [hunk 1]
```diff
@@ -149,3 +152,16 @@ func clusterChecksDisabledHandler(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusNotFound)
 	w.Write([]byte("Cluster-checks are not enabled"))
 }
+
+// parseClientIP retrieves the http client IP from the remoteAddr attribute
+func parseClientIP(remoteAddr string) string {
```
Comment (CharlyF) [inline cmd/cluster-agent/api/v1/clusterchecks.go]: Can we add unit tests to this ? I might be missing something, but if you are using IPV6, would that be supported ?

File: pkg/clusteragent/clusterchecks/dispatcher_main.go [hunk 2]
```diff
@@ -39,6 +44,19 @@ func newDispatcher() *dispatcher {
 		d.extraTags = append(d.extraTags, fmt.Sprintf("%s:%s", clusterTagName, clusterTagValue))
 	}
 
+	d.advancedDispatching = config.Datadog.GetBool("cluster_checks.advanced_dispatching_enabled")
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/dispatcher_main.go]: I think this is a good idea

File: pkg/clusteragent/clusterchecks/dispatcher_main.go [hunk 3]
```diff
@@ -39,6 +44,19 @@ func newDispatcher() *dispatcher {
...
+	if d.advancedDispatching {
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/dispatcher_main.go]: fast return here maybe ?
```
    d.clcRunnersClient = nil // might not even be necessary
	if !d.advancedDispatching {
      return d
    }
	var err error
    d.clcRunnersClient, err = clusteragent.GetCLCRunnerClient()
    if err != nil {
	  log.Warnf("Cannot create CLC runners client, advanced dispatching will be disabled: %v", err)
	  d.advancedDispatching = false
	}
	return d	
	}
```

File: pkg/clusteragent/clusterchecks/dispatcher_main.go [hunk 4]
```diff
@@ -161,6 +182,12 @@ func (d *dispatcher) run(ctx context.Context) {
 				danglingConfs := d.retrieveAndClearDangling()
 				d.reschedule(danglingConfs)
 			}
+		case <-ruunnerStatsTicker.C:
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/dispatcher_main.go]: s/ruunnerStatsTicker/runnerStatsTicker

File: pkg/clusteragent/clusterchecks/dispatcher_nodes.go [hunk 5]
```diff
@@ -122,3 +122,26 @@ func (d *dispatcher) expireNodes() {
 		log.Warn("No nodes reporting, cluster checks will not run")
 	}
 }
+
+// updateRunnersStats collects stats from the registred
+// Cluster Level Check runners and updates the stats cache
+func (d *dispatcher) updateRunnersStats() {
+	if d.clcRunnersClient == nil {
+		log.Debug("Couldn't instantiate a CLC Runner client")
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: This will be run every `runnerStatsTicker`, I think the message could be something like
`Cluster Level Check runner client was not correctly initialised` or something like that

File: pkg/clusteragent/clusterchecks/dispatcher_nodes.go [hunk 6]
```diff
@@ -122,3 +122,26 @@ func (d *dispatcher) expireNodes() {
...
+		return
+	}
+
+	d.store.Lock()
+	defer d.store.Unlock()
+	for name, node := range d.store.nodes {
+		node.Lock()
+		stats, err := d.clcRunnersClient.GetRunnerStats(node.clientIP)
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: Just curious about the timeout on this call. If it hangs, it seems like the struct will be locked. Could we lock, copy, unlock and then use the value ?

File: pkg/clusteragent/clusterchecks/stores.go [hunk 7]
```diff
@@ -52,12 +52,16 @@ func (s *clusterStore) getNodeStore(nodeName string) (*nodeStore, bool) {
 
 // getOrCreateNodeStore retrieves the store struct for a given node name.
 // If the node is not yet in the store, an entry will be inserted and returned.
-func (s *clusterStore) getOrCreateNodeStore(nodeName string) *nodeStore {
+func (s *clusterStore) getOrCreateNodeStore(nodeName, clientIP string) *nodeStore {
 	node, ok := s.nodes[nodeName]
 	if ok {
+		if node.clientIP != clientIP && clientIP != "" {
+			log.Debugf("Client IP changed for node %s: updating %s to %s", nodeName, node.clientIP, clientIP)
+			node.clientIP = clientIP
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/stores.go]: Do you need to `node.Lock` this ?

File: pkg/clusteragent/clusterchecks/types/types.go [hunk 8]
```diff
@@ -58,10 +58,11 @@ type Stats struct {
 // need and allows to inject a custom one for tests
 type LeaderIPCallback func() (string, error)
 
-// EndpointsInfo is used to store the collected info of endpoints.
```
Comment (CharlyF) [inline pkg/clusteragent/clusterchecks/types/types.go]: Why is this removed ?

File: pkg/config/config.go [hunk 9]
```diff
@@ -404,6 +404,8 @@ func initConfig(config Config) {
 	config.BindEnvAndSetDefault("cluster_checks.warmup_duration", 30)         // value in seconds
 	config.BindEnvAndSetDefault("cluster_checks.cluster_tag_name", "cluster_name")
 	config.BindEnvAndSetDefault("cluster_checks.extra_tags", []string{})
+	config.BindEnvAndSetDefault("cluster_checks.advanced_dispatching_enabled", false)
+	config.BindEnvAndSetDefault("cluster_checks.clc_runners_port", 5001)
```
Comment (CharlyF) [inline pkg/config/config.go]: why not using `cmd_port ` ?

File: pkg/util/clusteragent/clcrunner.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	// Set headers
+	c.clcRunnerAPIRequestHeaders = http.Header{}
+	c.clcRunnerAPIRequestHeaders.Set(authorizationHeaderKey, fmt.Sprintf("Bearer %s", authToken))
+
+	// Set http client
+	// TODO remove insecure
+	c.clcRunnerAPIClient = util.GetClient(false)
+	c.clcRunnerAPIClient.Timeout = 2 * time.Second
```
Comment (CharlyF) [inline pkg/util/clusteragent/clcrunner.go]: 👍

File: pkg/util/clusteragent/clcrunner.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+	// Set http port used by the CLC Runners
+	c.clcRunnerPort = config.Datadog.GetInt("cluster_checks.clc_runners_port")
+
+	return nil
+}
+
+// GetVersion fetches the version of the CLC Runner
+func (c *CLCRunnerClient) GetVersion(IP string) (version.Version, error) {
+	const clcRunnerVersionPath = "version"
```
Comment (CharlyF) [inline pkg/util/clusteragent/clcrunner.go]: worth moving with the other const at the top ?

File: pkg/util/clusteragent/clcrunner.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	err = json.Unmarshal(body, &version)
+
+	return version, err
+}
+
+// GetRunnerStats fetches the runner stats exposed by the Cluster Level Check Runner
+func (c *CLCRunnerClient) GetRunnerStats(IP string) (types.CLCRunnersStats, error) {
+	const clcRunnerStatsPath = "stats"
```
Comment (CharlyF) [inline pkg/util/clusteragent/clcrunner.go]: same as the "version" const

File: pkg/util/clusteragent/clcrunner_test.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+	require.Nil(suite.T(), err, fmt.Sprintf("%v", err))
+
+	expected := version.Version{
+		Major:  0,
+		Minor:  0,
+		Patch:  0,
+		Pre:    "test",
+		Meta:   "test",
+		Commit: "1337",
```
Comment (CharlyF) [inline pkg/util/clusteragent/clcrunner_test.go]: 🔥

File: cmd/cluster-agent/api/v1/clusterchecks.go (hunk 1 above)
Comment (ahmed-mez) [reply cmd/cluster-agent/api/v1/clusterchecks.go]: good catch 👌

File: pkg/clusteragent/clusterchecks/dispatcher_main.go (hunk 4 above)
Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/dispatcher_main.go]: thanks!

File: pkg/clusteragent/clusterchecks/dispatcher_nodes.go (hunk 5 above)
Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: makes sense 👍

File: pkg/clusteragent/clusterchecks/dispatcher_nodes.go (hunk 6 above)
Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: we timeout after 2 seconds  `c.clcRunnerAPIClient.Timeout = 2 * time.Second`, wdyt?

File: pkg/clusteragent/clusterchecks/stores.go (hunk 7 above)
Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/stores.go]: I don't think so, we already lock the `store` when calling this function, also we made it clean in the `clusterStore` and `nodeStore` struct definitions that only the dispatcher must hold the lock.

```
// nodeStore holds the state store for one node.
// Lock is to be held by the user (dispatcher)
type nodeStore struct {
...
```

```
// clusterStore holds the state of cluster-check management.
// Lock is to be held by the dispatcher so it can make atomic
// operations involving several calls.
type clusterStore struct {
...
```

File: pkg/clusteragent/clusterchecks/types/types.go (hunk 8 above)
Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/types/types.go]: the struct isn't used anymore, I should have cleaned it during the Endpoints check feature revamp

File: pkg/config/config.go (hunk 9 above)
Comment (ahmed-mez) [reply pkg/config/config.go]: indeed we can use `cmd_port` 👌

File: pkg/util/clusteragent/clcrunner.go (hunk 11 above)
Comment (ahmed-mez) [reply pkg/util/clusteragent/clcrunner.go]: tried to keep the same style we use in the DCA client in the same package

https://github.com/DataDog/datadog-agent/blob/master/pkg/util/clusteragent/clusteragent.go#L141-L143

https://github.com/DataDog/datadog-agent/blob/master/pkg/util/clusteragent/clusteragent.go#L200-L201

https://github.com/DataDog/datadog-agent/blob/master/pkg/util/clusteragent/clusteragent.go#L235-L236

**Comment by ahmed-mez** [comment]
@CharlyF Thank you for the review 💯 
I've made updates and I also added a small logic to calculate busyness based on the collected stats. The rebalancing algorithm will be implemented in another PR but based on this branch.

Looking forward to a 2nd review 🔥

File: pkg/clusteragent/clusterchecks/dispatcher_nodes.go (hunk 6 above)
Comment (CharlyF) [reply pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: Yes, I think we should copy otherwise we spend 2 seconds in deadlock

Comment (ahmed-mez) [reply pkg/clusteragent/clusterchecks/dispatcher_nodes.go]: done

**Review by CharlyF — APPROVED** [review]

**Comment by ahmed-mez** [comment]
replaced by https://github.com/DataDog/datadog-agent/pull/4068
